---
title: 'Class 15: RNASeq Analysis'
author: "Julia Napoli"
date: "11/17/2021"
output:
  pdf_document: default
  html_document: default
---

# Background

Today we're examining a published RNA-seq experiment where airway smooth muscle cells were treated with dexamethasone, a synthetic glucocorticoid steroid with anti-inflammatory effects (Himes et. al).

We need:
1) count data
2) col data

```{r}
counts <- read.csv("airway_scaledcounts.csv", row.names=1)
metadata <-  read.csv("airway_metadata.csv")
```

Let's take a look at each.
```{r}
head(counts)
head(metadata)
```

Let's check the correspondence of the metadata and count data setup.

```{r}
all(metadata$id == colnames(counts))
```

Let's perform some exploratory differential gene expression analysis.

## Compare control to treated

First we need to access all the control columns in our counts data.

```{r}
control.inds <- metadata$dex == 'control'
control.ids <- metadata[control.inds,]$id
```

Use these ids to access just the control columns of our `counts` data.

```{r}
control.mean <- rowMeans(counts[,control.ids])
head(control.mean)
```

Do the same for drug treated...

```{r}
drug.inds <- metadata$dex == 'treated'
drug.ids <- metadata[drug.inds,]$id
drug.mean <- rowMeans(counts[,drug.ids])
head(drug.mean)
```

We will combine our meancount data for bookkeeping purposes.

```{r}
meancounts <- data.frame(control.mean,drug.mean)
head(meancounts)
```

There are `r nrow(counts)` rows/genes in this dataset.

```{r}
nrow(counts)
```

## Compare the control and treated


```{r}
colSums(meancounts)
```

Let's visualize the results...

```{r}
plot(meancounts)
library(ggplot2)
ggplot(meancounts, aes(control.mean,drug.mean)) + geom_point()
```

This would benefit from a log transform! Let's do that.

```{r}
plot(meancounts, log = "xy")
```

We often use log transforms, bc they make life easier sometimes :)

```{r}
log2(20/20)
log2(40/20)
log2(10/20)
log2(80/20)
```

```{r}
meancounts$log2fc <- log2(meancounts[,"drug.mean"]/meancounts[,"control.mean"])
head(meancounts)
```

We need to drop the zero count genes/rows!

```{r}
head(meancounts[,1:2]==0)
```

The `which()` function tells us the indices of TRUE entries in a logical vector.

```{r}
which(c(T,F,T))
```

However, it is not that useful in default mode on our type of multi-column input.

```{r}
ind <- which(meancounts[,1:2] == 0, arr.ind = TRUE)
head(ind)
```

I only care about rows here (if there is a zero in any column I will exclude this row eventually).

```{r}
to.rm <- unique(sort(ind[,"row"]))
mycounts <- meancounts[-to.rm,]
```

We now have `r nrow(mycounts)` genes remaining.

```{r}
nrow(mycounts)
```
 
How many of these genes are up regulated at the log 2-fold change threshold of +2 or greater?

```{r}
sum(mycounts$log2fc > +2)
```

What percentage is this?

```{r}
round((sum(mycounts$log2fc > +2) / nrow(mycounts)*100),2)
```

How about down-regulated genes?

```{r}
sum(mycounts < -2)
```

## DESeq2 Analysis

```{r}
library(DESeq2)
```

We first need to steup the DESeq input object.

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex)
```

Run the DESeq analysis pipeline.

```{r}
dds <- DESeq(dds)
```

Look at the results.

```{r}
res <- results(dds)
head(res)
```

# Volcano plot

This is a very common data viz of this type of data that does not really look like a volcano.

```{r}
plot(res$log2FoldChange,res$padj)
plot(res$log2FoldChange,-log(res$padj), xlab="Log2(FoldChange)",
      ylab="-Log(P-value)")
abline(v=c(-2,2),col = "gray")
abline(h = -log(0.05),col ="gray")
```

## Adding annotation data

Let's add some meaningful gene names to our dataset so we can make sense of what is going on here!

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
```

```{r}
columns(org.Hs.eg.db)
```

Here we map to "SYMBOL" the common gene name that the world understands and wants.

```{r}
res$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),      # Our gene names
                     keytype="ENSEMBL",        # The format of our gene names
                     column="SYMBOL",          # The new format we want to add
                     multiVals="first")
head(res$symbol)
```


```{r}
head(res)
```

# Let's finally save our data to a file

```{r}
write.csv(res, file = "RNASeq_data")
```

# Pathway analysis

Let's try to bring some biology insight back into this work. For this we will start with KEGG.

```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)

# Examine the first 2 pathways in this kegg set for humans
head(kegg.sets.hs, 2)
```

Before we can use KEGG, we need to get our gene identifiers in the correct format for KEGG, which is ENTREZ format in this case.

```{r}
head(rownames(res))
```

```{r}
columns(org.Hs.eg.db)
```


```{r}
res$entrez <- mapIds(org.Hs.eg.db,
                     keys = row.names(res),
                     keytype = "ENSEMBL",
                     column = "ENTREZID",
                     multiVals = "first")

res$genename <- mapIds(org.Hs.eg.db,
                     keys = row.names(res),
                     keytype = "ENSEMBL",
                     column = "GENENAME",
                     multiVals = "first")

head(res)
```

The main gage() function requires a named vector of fold changes, where the names of the values are the Entrez gene IDs.

Note that we used the mapIDs() function above to obtain Entrez gene IDs (stored in res$entrez) and we have the fold change results from DESeq2 analysis (stored in res$log2FoldChange).

```{r}
foldchanges = res$log2FoldChange

names(foldchanges) = res$entrez
head(foldchanges)
```

```{r}
# Get the results

keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

We can look at the attributes() of this or indeed any R object

```{r}
attributes(keggres)
```

```{r}
head(keggres$less, 3)
```

The pathwview() function will add our genes to a KEGG pathway as colored entries: 

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05310")
```

![](hsa05310.pathview.png)


